modules/server/src/main/java/org/jboss/wsf/stack/cxf/deployment/WSDLFilePublisher.java

```diff
@@ -118,24 +119,61 @@ public void publishWsdlFiles(QName serviceName, String wsdlLocation, Bus bus, Li
 
    public void unpublishWsdlFiles()
    {
-      try
+      String deploymentName = dep.getCanonicalName();
+
+      if (deploymentName.startsWith("http://"))
       {
-         File parentDir = new File(serverConfig.getServerDataDir().getCanonicalPath() + "/wsdl");
-         ArchiveDeployment deployment = dep;
-         while (deployment.getParent() != null)
-         {
-            deployment = deployment.getParent();
-         }
-         String deploymentName = deployment.getCanonicalName();
-         if (deploymentName.startsWith("http://"))
-         {
-            deploymentName = deploymentName.replace("http://", "http-");
-         }
-         File targetDir = new File(parentDir, deploymentName);
-         FileUtils.removeDir(targetDir);
+         deploymentName = deploymentName.replace("http://", "http-");
       }
-      catch (IOException e)
-      {
+
+      try {
+
+         File publishDir = new File(serverConfig.getServerDataDir().getCanonicalPath()
+             + "/wsdl/" + deploymentName);
+
+         File wsdlRoot = new File(serverConfig.getServerDataDir().getCanonicalPath()
+             + "/wsdl/");
+         String wsdlRootDirStr = wsdlRoot.getCanonicalPath();
+
+         if (publishDir.exists()) {
+
+            List<File> dirList = new ArrayList<>();
+            getWsdlDirList(publishDir, dirList);
+            dirList.add(publishDir);  // clean up contents in the parent
+
+            for (File dir : dirList) {
+               File[] wsdlFileList = dir.listFiles();
+               if (wsdlFileList != null) {
+                  for (int i = 0; i < wsdlFileList.length; i++) {
+                     File f = wsdlFileList[i];
+                     FileUtils.delete(f);
+                     if (f.exists()) {
+                        Loggers.DEPLOYMENT_LOGGER.couldNotDeleteWsdlFile(f.getAbsolutePath());
+                     } else {
+                        Loggers.DEPLOYMENT_LOGGER.deletedWsdlFile(f.getAbsolutePath());
+                     }
+                  }
+               }
+            }
+
+            File parent = publishDir;
+            while(!wsdlRootDirStr.equals(parent.getCanonicalPath())) {
+
+               synchronized(lock) {
```

Comment:
I'm wondering... shouldn't the synchronized block contain the whole while loop here? IOW, shouldn't the removal of parent and grand-parents' (contents) be considered as an atomic operation?